### PR TITLE
Proposal: Graphsync Public Interfaces

### PR DESCRIPTION
**Pull request by hannahhoward**
# Goals

Outline public interfaces for using GraphSync for review by Filecoin stakeholders

I did this by completing an initial Readme documenting project usage so that once approved we'll also have documentation. Recommend choosing the [rich diff view](https://github.com/ipfs/go-graphsync/pull/20/files?short_path=04c6e90#diff-04c6e90faac2675aa89e2176d2eec7d8) when reviewing this PR

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,205 @@
...
+rootedSelector := graphsync.SelectorSpecFromCidAndPath(lnk, pathSegments)
+```
+
+### Response Type
+
+```golang
+
+type ResponseProgress struct {
+  Node ipld.Node // a node which matched the graphsync query
+  Path      ipld.Path // the path of that node relative to the traversal start
```
Comment (warpfork) [inline README.md]: odd whitespace?

**Comment by warpfork** [comment]
awesome docsing :+1: :+1: :+1:

File: README.md [hunk 2]
```diff
@@ -1 +1,205 @@
...
+A rooted selector is a `go-ipld-prime` node that follows the spec outlined here: https://github.com/ipld/specs/blob/master/selectors/selectors.md
+
+`go-ipld-prime` provides a series of builder interfaces for building this kind of structured data into a node. If your library simply wants to make a selector from CID and a path, represented by an array of strings, you could construct the node as follows:
+
+```golang
+import (
+	ipld "github.com/ipld/go-ipld-prime"
+	free "github.com/ipld/go-ipld-prime/impl/free"
+	fluent "github.com/ipld/go-ipld-prime/fluent"
+  cidLink "github.com/ipld/go-ipld-prime/linking/cid"
```
Comment (lanzafame) [inline README.md]:
```suggestion
	cidLink "github.com/ipld/go-ipld-prime/linking/cid"
```

**Comment by lanzafame** [comment]
@hannahhoward the title of the PR says Graphsync Public Interfaces, but I feel that the interface may have been drowned out by the `ipld-prime` docs. It may be worth clearly outlining the interface by itself.

File: README.md [hunk 3]
```diff
@@ -1 +1,205 @@
...
+
+## Contribute
+
+PRs are welcome!
+
+Small note: If editing the Readme, please conform to the [standard-readme](https://github.com/RichardLitt/standard-readme) specification.
+
+## License
+
+MIT © Protocol Labs
```
Comment (momack2) [inline README.md]: I believe we are actually a dual MIT & Apache-2 License re https://github.com/ipfs/team-mgmt/issues/849

Comment (hannahhoward) [reply README.md]: Will do.

BTW, that was a copy from the Bitswap readme so we should probably update that as well.

**Comment by hannahhoward** [comment]
@lanzafame I agree there's a lot of "oh and here's some go-ipld-prime stuff you need to understand to make this work". I will probably separate that out into a separate compatibility doc at some point. But for now I'm inclined to present it together since it's relevant to all the early clients of the library just to get up and running kind of.

File: README.md (hunk 3 above)
Comment (hannahhoward) [reply README.md]: Went ahead and updated and also added both licenses in master (didn't want to pollute this PR with more stuff)

**Comment by hannahhoward** [comment]
Eek... I messed up something. Closing for a few minutes in case anyone is reading.

**Comment by hannahhoward** [comment]
Ok, sorry, problem solved. Change to initialization steps, FYI.

**Pull request merged**